### PR TITLE
Use FeatureBindings to fold web sockets into the normal flow

### PR DESCRIPTION
**Pull request by swankjesse**
This makes it so that web sockets participate in interceptors.

### CODE REVIEW

File: misk/src/main/kotlin/misk/web/jetty/JettyServletUpstreamResponse.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package misk.web.jetty
+
+import misk.web.ServletHttpCall
+import misk.web.actions.WebSocketListener
+import okhttp3.Headers
+import org.eclipse.jetty.http.HttpFields
+import org.eclipse.jetty.server.Response
+import java.util.function.Supplier
+
+internal class JettyServletUpstreamResponse(
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/jetty/JettyServletUpstreamResponse.kt]: this code is moved from WebActionsServlet.kt, line 120

File: misk/src/main/kotlin/misk/web/jetty/JettyWebSocket.kt [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import org.eclipse.jetty.websocket.api.WebSocketAdapter
+import org.eclipse.jetty.websocket.api.WriteCallback
+import org.eclipse.jetty.websocket.servlet.ServletUpgradeRequest
+import org.eclipse.jetty.websocket.servlet.ServletUpgradeResponse
+import org.eclipse.jetty.websocket.servlet.WebSocketCreator
+import java.util.ArrayDeque
+
+private const val MAX_QUEUE_SIZE = 16 * 1024 * 1024
+
+internal class JettyWebSocket(
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/jetty/JettyWebSocket.kt]: Mostly moved from RealWebSocket.kt and WebActionsServlet

File: misk/src/main/kotlin/misk/web/BoundAction.kt [hunk 3]
```diff
@@ -87,18 +104,6 @@ internal class BoundAction<A : WebAction>(
     chain.proceed(httpCall)
   }
 
-  internal fun handleWebSocket(
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/BoundAction.kt]: one fewer branch!

File: misk/src/main/kotlin/misk/web/BoundAction.kt [hunk 4]
```diff
@@ -167,21 +172,11 @@ internal open class RequestMatch(
...
   requestCharsetMatch: Boolean,
   responseContentType: MediaType
-) : RequestMatch(action.pathPattern, acceptedMediaRange, requestCharsetMatch, responseContentType) {
-
-  /** Handles [httpCall] by handing it off to the action. */
-  fun handle(httpCall: HttpCall) {
-    action.handle(httpCall, pathMatcher)
-  }
-
-  fun handleWebSocket(httpCall: HttpCall): WebSocketListener {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/BoundAction.kt]: one fewer branch!

File: misk/src/main/kotlin/misk/web/ServletHttpCall.kt [hunk 5]
```diff
@@ -77,6 +78,10 @@ internal data class ServletHttpCall(
     this.webSocket = webSocket
   }
 
+  override fun initWebSocketListener(webSocketListener: WebSocketListener) {
+    this.upstreamResponse.initWebSocketListener(webSocketListener)
```
Comment (oldergod) [inline misk/src/main/kotlin/misk/web/ServletHttpCall.kt]: The `FakeHttpCall` has a validation that init is being called only once. 
```check(this.webSocketListener == null) { "web socket listener already set" }```

Should we put one here?

File: misk/src/main/kotlin/misk/web/jetty/JettyWebSocket.kt [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        response.addHeader(headers.name(i), headers.value(i))
+      }
+    }
+
+    override fun requireTrailers() = error("no trailers for web sockets")
+
+    override fun setTrailer(name: String, value: String) = error("no trailers for web sockets")
+
+    override fun initWebSocketListener(webSocketListener: WebSocketListener) {
+      this@JettyWebSocket.listener = webSocketListener
```
Comment (oldergod) [inline misk/src/main/kotlin/misk/web/jetty/JettyWebSocket.kt]: Again, do we want a
```check(this@JettyWebSocket.listener == null) { "web socket listener already set" }```
?

**Comment by swankjesse** [comment]
Will address comments in follow-up

**Pull request merged**